pkg/datapath/linux/devices.go

```diff
@@ -278,15 +288,139 @@ func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.
 
 		viable := dm.isViableDevice(l3DevOK, info.hasDefaultRoute, link)
 		if viable {
+			log.WithField(logfields.Device, name).Info("Added new device")
 			dm.devices[name] = struct{}{}
 			changed = true
+		} else {
+			log.WithField(logfields.Device, name).Debug("Skipping unviable device")
 		}
 	}
 	return changed
 }
 
+func (dm *DeviceManager) handleLinkUpdate(update netlink.LinkUpdate) bool {
+	if update.Header.Type != unix.RTM_DELLINK {
+		return false
+	}
+
+	name := update.Attrs().Name
+
+	if _, ok := dm.devices[name]; !ok {
+		return false
+	}
+
+	delete(dm.devices, name)
+
+	return true
+}
+
+// Listen starts listening to changes to network devices. When devices change the new set
+// of devices is sent on the returned channel.
+func (dm *DeviceManager) Listen(ctx context.Context) (chan []string, error) {
+	return dm.listen(ctx, nil)
+}
+
+// listen is the internal method to start listening for changes to network devices. This is split from the
+// public method to allow tests to listen to devices in specific network namespace.
+func (dm *DeviceManager) listen(ctx context.Context, netNS *netns.NsHandle) (chan []string, error) {
+	l3DevOK := supportL3Dev()
+	routeChan := make(chan netlink.RouteUpdate)
+	closeChan := make(chan struct{})
+
+	err := netlink.RouteSubscribeWithOptions(routeChan, closeChan,
+		netlink.RouteSubscribeOptions{
+			Namespace: netNS,
+			// List existing routes to make sure we did not miss changes that happened after Detect().
+			// FIXME(JM): Perhaps could consider combining Detect() and Listen() so we don't have to do this.
+			// Having this whole thing be fully event-driven is likely a better design as that forces consumers
+			// to work with it correctly.
+			ListExisting: true,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	linkChan := make(chan netlink.LinkUpdate)
+	err = netlink.LinkSubscribeWithOptions(linkChan, closeChan, netlink.LinkSubscribeOptions{
+		Namespace: netNS,
+		// FIXME(JM): What if a device is removed between Detect() and Listen()?
+		ListExisting: false,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	devicesChan := make(chan []string)
+
+	go func() {
+		// If a specific namespace is requested, lock the thread and set the
+		// threads namespace. This is currently only relevant for testing.
+		if netNS != nil {
+			runtime.LockOSThread()
+			defer runtime.UnlockOSThread()
```

Comment:
Mildly interesting point, I guess it probably doesn't matter here given the usage for tests, but this `defer()` will only be run when the goroutine exits, which would only happen once the context is cancelled.. not sure I understand the implications here given that the runtime needs to ensure that at least one thread is available in this namespace the whole time, so that this goroutine can execute in that namespace.

I've seen this pattern before with syscalls, but not so much to bound long-living goroutines.

(No action necessary, just pondering whether this is a pattern we should be vigilant about if it ever gets introduced outside test code).

Reply:
This has been now removed. I instead opted for using `netlink.NewHandleAt` for the netlink operations so no need for the thread pinning.